### PR TITLE
[Hunyuan]: add Hunyuan Dense V1 causal language model support

### PR DESCRIPTION
**Pull request by KissrainKlein**
本 PR 完成 Hunyuan Dense V1 从 Transformers 到 PaddleFormers 的迁移，包含组网、前向精度、前 10 token 生成对齐、GSM8K 训练 loss 曲线对比和模型单测验证。

### 1. 模型实现与注册

- 新增 Hunyuan Dense V1 模型、配置和 tokenizer：
  - `paddleformers/transformers/hunyuan/configuration.py`
  - `paddleformers/transformers/hunyuan/modeling.py`
  - `paddleformers/transformers/hunyuan/tokenizer.py`
  - `paddleformers/transformers/hunyuan/tokenizer_fast.py`
- 支持 `hunyuan_v1_dense` 配置识别，并完成 `AutoConfig`、`AutoModel`、`AutoModelForCausalLM` 注册。
- 实现 Hunyuan GQA、DynamicNTK RoPE、KV cache、CausalLM、序列分类与 token 分类路径。

### 2. 前向精度对齐

- 模型：`Hunyuan-0.5B-Instruct`
- 使用 5 个中英文 prompt 对比 Transformers 与 PaddleFormers 的前向 logits。
- 5/5 个样例 logits 均满足 `rtol=1e-3, atol=1e-2` 的 `allclose`。
- 最大 logits 绝对误差：`2.6464462e-05`
- 各样例 mean absolute difference 的平均值：`1.9950181e-06`

### 3. 生成前向进度对齐

相同权重、greedy decoding（`do_sample=False`）和 `max_new_tokens=32` 下，使用 5 个中英文 prompt 验证生成前向进度：

- 前 10 个生成 token IDs 完全一致。
case1：
HF first 10 new token ids: [120029, 185, 3125, 270, 3890, 558, 14746, 507, 17687, 2842]
Paddle first 10 new token ids: [120029, 185, 3125, 270, 3890, 558, 14746, 507, 17687, 2842]
case2：
HF first 10 new token ids: [120029, 185, 3125, 270, 3890, 14116, 58131, 1668, 292, 2733]
Paddle first 10 new token ids: [120029, 185, 3125, 270, 3890, 14116, 58131, 1668, 292, 2733]
case3：
HF first 10 new token ids: [120029, 185, 56520, 11, 252, 2388, 316, 14053, 356, 252]
Paddle first 10 new token ids: [120029, 185, 56520, 11, 252, 2388, 316, 14053, 356, 252]
case4：
HF first 10 new token ids: [120029, 185, 3890, 3030, 1115, 2150, 507, 16, 10, 16]
Paddle first 10 new token ids: [120029, 185, 3890, 3030, 1115, 2150, 507, 16, 10, 16]
case5：
HF first 10 new token ids: [120029, 185, 3125, 270, 3890, 14116, 1824, 507, 7331, 12523]
Paddle first 10 new token ids: [120029, 185, 3125, 270, 3890, 14116, 1824, 507, 7331, 12523]

### 4. 训练 loss 曲线对比

- 数据：固定顺序 GSM8K canonical JSONL。
- 训练：SFT、full finetuning、每卡 batch size=1、gradient accumulation=4、300 optimizer steps。
- 对比对象：PaddleFormers 与 ms-swift；两端学习率合同已逐 step 核验。
- 两端均拥有连续的 300 个 step 记录，所有 loss、学习率和差值均为有限数，无 NaN/Inf。
- 最大学习率绝对差：`4.2857143e-10`。

| 指标 | PaddleFormers | ms-swift / 对比结果 |
| --- | ---: | ---: |
| 第 1 步 loss | `2.3506793126` | `2.3583018780` |
| 第 300 步 loss | `0.7751995468` | `0.9477578998` |
| 300-step 平均绝对 loss 差 | - | `0.1626918178` |

### 5. 模型单测

新增 `tests/transformers/hunyuan/test_modeling.py`，覆盖：

- 基础前向和 attention mask；
- CausalLM logits 与带 labels 的 loss；
- 显式/隐式 position IDs；
- KV cache 增量解码与完整前向最后一个 token logits；
- GQA；
- `input_ids` / `inputs_embeds` 等价前向；
- 序列分类和 token 分类；
- 通用 mixin 的配置序列化、forward signature、save/load、determinism、embedding resize 和 generation。

### CODE REVIEW

**Comment by a31413510** [comment]
检查ci单侧报错
image

**Comment by a31413510** [comment]
建议在PaddleFormers/paddleformers/datasets/template/template.py中增加hunyuan模版，可参考：
register_template(
      name="hunyuan",
      format_user=StringFormatter(slots=["<｜hy_User｜>{{content}}<｜hy_Assistant｜>"]),
      format_assistant=StringFormatter(slots=["{{content}}", {"eos_token"}]),
      format_prefix=EmptyFormatter(slots=[{"bos_token"}]),
  )

**Comment by a31413510** [comment]
paddleformers/transformers/glm4/modeling.py和tests/transformers/glm4/test_glm4.py中的所有中文注释都需要改成英文或者直接删除

File: paddleformers/transformers/hunyuan/modeling.py [hunk 1]
```diff
@@ -0,0 +1,1257 @@
...
+        gpt_model._gen_inv_aoa_config = cls._gen_inv_aoa_config
+        if not hasattr(config, "architectures"):
+            config.architectures = [cls.__name__.replace("Pipe", "")]
+        gpt_model.config_to_save = config
+        gpt_model.is_fleet = cls.is_fleet
+        return gpt_model
+
+
+# Exact Hugging Face architecture name used by Hunyuan Dense V1 checkpoints.
+HunYuanDenseV1ForCausalLM = HunyuanForCausalLM
```
Comment (risemeup1111) [inline paddleformers/transformers/hunyuan/modeling.py:1274]: P1 **优先级：P1**
这里目前只补了 `HunYuanDenseV1ForCausalLM`，但 `auto/modeling.py` 会按最长前缀把 `HunYuanDenseV1ForCausalLM` 解析成 `HunYuanDenseV1Model`。现在这个别名不存在，所以 `AutoModel.from_pretrained()` 遇到 Hunyuan Dense V1 checkpoint 时会直接找不到类。**处理要求：请针对该评论修复并提交新的 commit。**

建议至少补一个：
```python
HunYuanDenseV1Model = HunyuanModel
```
如果还要支持 pipe / 任务头，再在对应类定义后补同名 alias。

File: paddleformers/transformers/__init__.py [hunk 2]
```diff
@@ -180,6 +180,16 @@
...
         "LinearAnnealingWithWarmupDecay",
     ],
+    "hunyuan.configuration": ["HunyuanConfig"],
+    "hunyuan.modeling": [
+        "HunyuanModel",
+        "HunyuanPretrainedModel",
+        "HunyuanForCausalLM",
+        "HunYuanDenseV1ForCausalLM",
+        "HunyuanForCausalLMPipe",
+    ],
```
Comment (risemeup1111) [inline paddleformers/transformers/__init__.py:217]: P2 **优先级：P2**
这里的顶层 lazy import 还是只暴露了基础模型和 CausalLM，`HunyuanForSequenceClassification` / `HunyuanForTokenClassification` / `HunyuanSentenceEmbedding` 仍然无法从 `paddleformers.transformers` 直接导入。既然这次 PR 已经新增了这些类，建议把它们也加进顶层导出；`paddleformers/transformers/hunyuan/__init__.py` 里最好同步一份。**处理要求：请针对该评论进行回复（同意并已修改请回复 Done，不同意请说明理由）。**

```suggestion
    "hunyuan.modeling": [
        "HunyuanModel",
        "HunyuanPretrainedModel",
        "HunyuanForCausalLM",
        "HunYuanDenseV1ForCausalLM",
        "HunyuanForCausalLMPipe",
        "HunyuanForSequenceClassification",
        "HunyuanForTokenClassification",
        "HunyuanSentenceEmbedding",
        "HunyuanForCausalLMFleet",
        "HunyuanForCausalLMPipeFleet",
    ],
```

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
已复查，当前还有需要修正的阻塞点，细节都在行内评论里。另有一条非行级的范围建议：
- P3 **优先级：P3** 非行级：`glm4` 目录也被带进了这次 diff，但 `auto/configuration.py` / `auto/modeling.py` / 顶层导出还没接上，和当前标题/描述的 Hunyuan 范围也不一致。请确认这部分是否要一起合入；如果要合入，请补齐注册和说明，否则建议拆到独立 PR。**处理要求：请针对该评论进行回复（同意并已修改请回复 Done，不同意请说明理由）。**
<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Comment by a31413510** [comment]
需要解决代码冲突
image

**Comment by a31413510** [comment]
需要解决代码冲突

File: paddleformers/transformers/hunyuan/modeling.py [hunk 3]
```diff
@@ -0,0 +1,1257 @@
...
+
+
+class HunyuanForCausalLMPipe(HunyuanForCausalLM):
+    config_class = HunyuanConfig
+    _decoder_layer_cls = HunyuanDecoderLayer
+    _get_tensor_parallel_mappings = HunyuanModel._get_tensor_parallel_mappings
```
Comment (risemeup1111) [inline paddleformers/transformers/hunyuan/modeling.py:1280]: P0 **优先级：P0**

这里在模块导入阶段读取 `HunyuanModel._get_tensor_parallel_mappings`，但该方法既没有在 Hunyuan 模型中定义，也不在 `PretrainedModel` 基类中提供。因此加载 `paddleformers.transformers.hunyuan.modeling`（包括顶层 lazy import 首次访问 Hunyuan）会在类定义时直接抛出 `AttributeError`，整个 Hunyuan 实现无法导入。请为 HunyuanModel 实现实际的 tensor-parallel mapping（并覆盖所需权重布局），或移除该引用并提供正确的 pipe 映射。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
| 序号 | 位置 | 优先级 | 状态 |
| --- | --- | --- | --- |
| 1 | [hunyuan/modeling.py:1236](https://github.com/PaddlePaddle/PaddleFormers/pull/4767#discussion_r3810287096) | P0 | 🚧 |
| 2 | [hunyuan/modeling.py:1230](https://github.com/PaddlePaddle/PaddleFormers/pull/4767#discussion_r3602248591) | P1 | 🚧 |
| 3 | [transformers/__init__.py:211](https://github.com/PaddlePaddle/PaddleFormers/pull/4767#discussion_r3602250576) | P2 | 🚧 |

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>
